VRF P256 Implementation Has a Flaw

Running the `vrf/p256/p256_test.go` 10K times:
- `TestVRF` fails ~1.2% of the times.
- `TestReadFromOpenSSL` fails ~0.35% of the times.

Sample output:

```
--- FAIL: TestVRF (0.14s)
    p256_test.go:75: 2: Verify([100 97 116 97 50], [4 101 217 159 12 203 153 201 56 199 77 58 145 45 164 130 220 170 142 106 242 209 229 149 163 63 89 45 39 24 35 205 43 130 250 91 252 159 32 82 123 182 105 227 165 93 251 65 116 110 87 1 19 102 51 155 233 218 238 83 68 204 227 130 157], [0 224 154 241 94 7 73 23 89 160 78 98 204 247 179 126 25 201 241 219 185 34 167 24 127 229 37 244 33 52 248 220 124 211 245 231 17 236 19 70 138 133 75 129 41 194 169 51 56 125 76 58 196 251 166 30 165 167 152 118 38 28 132 73]): got invalid VRF proof, want <nil>
```

```
--- FAIL: TestReadFromOpenSSL (0.04s)
    p256_test.go:129: Failed verifying VRF proof
```

This is no longer caused by #243.

--- a/core/vrf/p256/p256.go
+++ b/core/vrf/p256/p256.go
@@ -161,7 +161,7 @@ func (pk *PublicKey) Verify(m, vrf, proof []byte) error {
 		return ErrInvalidVRF
 	}
 
-	// Parse proof into s and t
+	// Parse proof into s and t.
 	s := proof[0:32]
 	t := proof[32:64]
 
@@ -186,7 +186,13 @@ func (pk *PublicKey) Verify(m, vrf, proof []byte) error {
 	b.Write(elliptic.Marshal(curve, h1TKSx, h1TKSy))
 	h2 := H2(b.Bytes())
 
-	if !hmac.Equal(s, h2.Bytes()) {
+	// Left pad h2 with zeros if needed. This will ensure that h2 is padded
+	// the same way s is.
+	var buf bytes.Buffer
+	buf.Write(make([]byte, 32-len(h2.Bytes())))
+	buf.Write(h2.Bytes())
+
+	if !hmac.Equal(s, buf.Bytes()) {
 		return ErrInvalidVRF
 	}
 	return nil

